Add 'diffbuf' option to only compare with a single buffer for highlighting

flarn2006 opened the pull request:
I needed this feature at work today, and it seemed like it would be simple enough to add, so after I got home I decided to give it a shot.

Basically, what this does is add a new option called `diffbuf` (short name `dbuf`) which defaults to 0 (meaning standard behavior) but can also be set to a buffer number. If it's set to a valid buffer number, then instead of highlighting all the lines that any buffer has different, it will only highlight lines that **that specific buffer** has different. That is, for the purpose of highlighting lines, it will be as if the buffer whose lines are to be highlighted and the buffer specified by `diffbuf` are the only buffers open; others won't affect it, beyond adding filler lines if that option is turned on.

Lines *inside* the buffer specified by `diffbuf` will be highlighted as normal.

This is my first contribution to this project—or any project as popular as Vim, for that matter—so please forgive me if I did anything incorrectly.

**EDIT:** I've taken some screenshots to show what the new feature does. Note that **except in the first screenshot, lines 15 and 16 are not representative of intended behavior**; this is a bug that I will explain shortly. So ignore those two lines if you're just trying to understand what this is supposed to do.

![diffbuf=0](https://user-images.githubusercontent.com/687313/46773948-02f8f700-ccce-11e8-9b2d-6e37bbe09178.png)

^ (`diffbuf=0`) This is the standard behavior, the way lines have always been highlighted in diff mode. That is, unless a line is the same in **every** diff buffer that's open, that line will show up as highlighted in **all** of the buffers. Notice how every line that differs in even one window is highlighted across the board. Zero is the default value of `diffbuf`, so the default behavior will not change.

![diffbuf=1](https://user-images.githubusercontent.com/687313/46773955-0a200500-ccce-11e8-806e-b9ba357aeda4.png)

^ (`diffbuf=1`) With `diffbuf` set to 1, however, it will only highlight differences with buffer 1. (The exception is buffer 1 itself, which is highlighted like before.) Line 5 is still highlighted in all rows because both of the other buffers have something different there. But line 10 is not highlighted in buffer 2, because there is no difference when it only compares against buffer 1. Buffer 3 says something different, and it's highlighted there, but with `diffbuf` set to anything but 0 (compare against all buffers) or 3 (compare against buffer 3) that doesn't matter. In other words, buffer 1 is the "master buffer", and if you're looking at any other buffer, it's only the master buffer that needs to be considered. **Remember, ignore lines 15 and 16 here. They aren't supposed to be like that.**

![diffbuf=2](https://user-images.githubusercontent.com/687313/46775686-3ccdfb80-ccd6-11e8-971e-fa584340f9d0.png)

^ (`diffbuf=2`) Here it is comparing against buffer 2 instead. Now buffer 2 is the one that's highlighted normally, and the other buffers are only highlighted if that line differs from buffer 2.

![diffbuf=3](https://user-images.githubusercontent.com/687313/46775708-58390680-ccd6-11e8-888a-d4d97917871e.png)

^ (`diffbuf=3`) Here's 3, not much more to say really.

## Known Issue

As you can see, this currently doesn't work properly with consecutive lines. Look at the third image, where `diffbuf` is 2, for instance. Lines 15 and 16 each differ from the middle buffer on one side, but not the other, but 16 differs on the left and 15 differs on the right. The proper behavior here would be to have line 15 only highlighted on the middle and right buffers, and line 16 only on the middle and left. But for some reason, it's treating consecutive lines like a block, and isn't highlighting only the ones that actually differ. If I were to `set nodiff` in the buffer on the right, line 15 will no longer be highlighted.

With the code working properly, however, changing something in a buffer other than the master one will not affect the appearance of any other buffers, beyond adding filler lines if necessary to keep everything together.

flarn2006 committed:
feat: add 'diffbuf' option (does nothing yet)

flarn2006 committed:
feat: implement diffbuf option

flarn2006 committed:
test: add 'diffbuf' to expected options output

k-takata commented:
There is a reply from @h-east at here: https://groups.google.com/d/msg/vim_dev/9CwgjallJDg/4A8Gi7xVBgAJ

flarn2006 committed:
style: make indentation consistent

flarn2006 committed:
fix: move var declarations to beginning

For some reason one of the appveyor builds fails otherwise. I thought
this was only an issue in obsolete C compilers but I guess not.

flarn2006 commented:
@k-takata, I replied to you on Google Groups; just making sure you saw.

Also, I noticed a case just now where it isn't working properly. But I'm at work so I'll look into fixing that when I get home.

brammool commented:
Please add a screenshot or two, so that we can see how this works.  Also, please add the help for this option, which a good explanation (can't add a screenshot in the help).

flarn2006 commented:
@brammool Will do. I'll also take a look at the part where it wasn't working and try and fix that.

flarn2006 commented:
Screenshots added. Will add help next.

Also, Gary Johnson posted in the Google Group:

> If this turns out to be something useful and worth including in Vim,
> I think we really need to come up with some better method of
> invoking it than setting the value of an option to a buffer number.
> 
> I don't understand what this does well enough to suggest a name, but
> I think some command like :diffref that sets the current buffer as
> the reference buffer, or whatever 'diffbuf' means, would be much
> better than requiring the user to figure out the number of the
> buffer of interest, then setting some option to it.  Further, an
> option value will hang around and confuse the user the next time
> they execute :diffthis.

That's a good point. Maybe keep the option though, and just add a command to set it to the current buffer, because potentially it can also be extended in the future to support a list of multiple buffers to compare against.

Actually, I think I'm going to change it to be a per-buffer option, `diffignore`/`nodiffignore` or something, defaulting to `nodiffignore`. Or alternatively `diffref`/`nodiffref`, defaulting to `diffref`. Probably in a different pull request come to think of it; I'll see what everyone thinks of that idea and then close this one if it's well-received.

brammool commented:
> Screenshots added. Will add help next.

Thanks, that makes it much clearer.

> Also, Gary Johnson posted in the Google Group:
> 
> > If this turns out to be something useful and worth including in Vim,
> > I think we really need to come up with some better method of
> > invoking it than setting the value of an option to a buffer number.
> > 
> > I don't understand what this does well enough to suggest a name, but
> > I think some command like :diffref that sets the current buffer as
> > the reference buffer, or whatever 'diffbuf' means, would be much
> > better than requiring the user to figure out the number of the
> > buffer of interest, then setting some option to it.  Further, an
> > option value will hang around and confuse the user the next time
> > they execute :diffthis.
> 
> That's a good point. Maybe keep the option though, and just add a command to set it to the current buffer, because potentially it can also be extended in the future to support a list of multiple buffers to compare against.
> 
> Actually, I think I'm going to change it to be a per-buffer option,
> `diffignore`/`nodiffignore` or something, defaulting to
> `nodiffignore`. Or alternatively `diffref`/`nodiffref`, defaulting to
> `diffref`.

Not sure if this is useful.  I think what most users will want is
setting one buffer as the one to diff all other buffers against.  For
example, you have the original file, your version and the version you
need to merge in.

How about adding a buffer-local option "diffmaster" that can be set.
You set it on the original file, so you get your changes highlighted and
you get the merge changes highlighted, but not the difference between
your changes and the merge changes.

When you switch it on, it will be switched off in all other buffers in
the current tab.  There can only be one master!

-- 
A bad peace is better than a good war. - Yiddish Proverb

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

flarn2006 commented:
Sure, so basically `diffref` from my suggestion above only it automatically turns off on all other other buffers when you set it on one?

I'll create a separate pull request when I'm ready; let me know if that sounds good.

vim-ml commented:
On 2018-10-11, flarn2006 wrote:
> Sure, so basically diffref from my suggestion above only it automatically turns
> off on all other other buffers when you set it on one?
> 
> I'll create a separate pull request when I'm ready; let me know if that sounds
> good.

So the value in all buffers is initially one, but [re]setting it to
one in one buffer seta all the others to zero?  Can we think about
this a little more?  Maybe a command to clear (i.e., set to zero)
the variable/option in all buffers but the current one?

Also, shouldn't this be per window rather than per buffer?

Regards,
Gary

Houl commented:
Reminds me of an old feature request: [Humble feature request: 3way-merge diff mode ](http://vim.1045645.n5.nabble.com/Humble-feature-request-3way-merge-diff-mode-td1204381.html).

brammool commented:
> On 2018-10-11, flarn2006 wrote:
> > Sure, so basically diffref from my suggestion above only it automatically turns
> > off on all other other buffers when you set it on one?
> > 
> > I'll create a separate pull request when I'm ready; let me know if that sounds
> > good.
> 
> So the value in all buffers is initially one, but [re]setting it to
> one in one buffer seta all the others to zero?  Can we think about
> this a little more?  Maybe a command to clear (i.e., set to zero)
> the variable/option in all buffers but the current one?
> 
> Also, shouldn't this be per window rather than per buffer?

Since 'diff' is per window, yes, this option would also be per window.

'diffmaster' would be default off, and it can be on for one window only.

-- 
A computer program does what you tell it to do, not what you want it to do.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

flarn2006 commented:
How about I do it like I said before, calling it `diffref` instead of `diffmaster` and having it be on by default and settable on multiple buffers at once, but then I also add `diffmaster` as a command that turns it on for the current buffer and off for all others? That way it's just as easy as before to use it for that most common use case of only comparing with a single buffer, but at the same time if someone does ever get the need to turn it on for multiple (but not all) buffers, they can do that too? From my understanding of the way it's coded, it would be no more difficult than only allowing a single master (actually less difficult because I won't need to have it check other buffers when you set the variable) so it would be a shame to explicitly remove that functionality if there's another way to get the ease of use it would bring.

Also you know the code a lot more than I do, but I'm pretty sure it actually would be a buffer option, not a window option. It's the buffer that it compares against. If two windows have the same buffer, what difference would it make which of the two is set as the master? The `diff` option is window-specific because it's specific to the way that window is treated, but the `diffref` or `diffmaster` option, whatever it's called, doesn't really involve the window; it just tells other windows whether or not to reference the buffer inside.

Btw, should I be concerned about that coverage check that's failing? I have no experience at all writing automated tests and wouldn't know where to begin.

vim-ml commented:
On 2018-10-11, flarn2006 wrote:

> Also you know the code a lot more than I do, but I'm pretty sure it actually
> would be a buffer option, not a window option. It's the buffer that it compares
> against. If two windows have the same buffer, what difference would it make
> which of the two is set as the master? The diff option is window-specific
> because it's specific to the way that window is treated, but the diffref or
> diffmaster option, whatever it's called, doesn't really involve the window; it
> just tells other windows whether or not to reference the buffer inside.

Consider, for example, the case of one buffer being opened in two
windows, each window in a different tab.  Each of those windows can
be diffed against any other window(s) in the same tab.  A buffer
could be displayed in the master window in one tab and in
a non-master window in another tab.

Regards,
Gary

flarn2006 commented:
Ah, good point. Window option it is. I'm going to close this pull request and open a new one when it's ready.

flarn2006 commented:
Since I'm not going to use this code (and shouldn't have used the master branch for it) I'm going to change history in my fork and put the master branch back to the official one. I've created a separate branch in my fork, [pullreq3527](https://github.com/flarn2006/vim/tree/pullreq3527), to hold these changes that aren't going to be used, since I'm not sure how this will affect the pull request.

brammool commented:
> How about I do it like I said before, calling it `diffref` instead of
> `diffmaster` and having it be on by default and settable on multiple
> buffers at once, but then I also add `diffmaster` as a command that
> turns it on for the current buffer and off for all others? That way it's
> just as easy as before to use it for that most common use case of only
> comparing with a single buffer, but at the same time if someone does
> ever get the need to turn it on for multiple (but not all) buffers, they
> can do that too? From my understanding of the way it's coded, it would
> be no more difficult than only allowing a single master (actually less
> difficult because I won't need to have it check other buffers when you
> set the variable) so it would be a shame to explicitly remove that
> functionality if there's another way to get the ease of use it would
> bring.

I have a hard time imagining how it would be used, have 'diffref' in
one window/buffer to be different from another one.  What are you
looking at then?

> Also you know the code a lot more than I do, but I'm pretty sure it
> actually would be a buffer option, not a window option. It's the
> buffer that it compares against. If two windows have the same buffer,
> what difference would it make which of the two is set as the master?
> The `diff` option is window-specific because it's specific to the way
> that window is treated, but the `diffref` or `diffmaster` option,
> whatever it's called, doesn't really involve the window; it just tells
> other windows whether or not to reference the buffer inside.

One of the most common things to do is the disable 'diff' and switch it
on in another window.  What would then happen with the buffer-local
option?  Having 'diffref' set in a buffer that doesn't appear in any
window with 'diff' set is like a misconfiguration.  We should try to
avoid that from happening.  Also, setting it for a hidden buffer doesn't
make sense.

It's true that setting the option really means to use the buffer in that
window as the master.  I don't see a problem with that.

Gary's remark also mentions a situation where a window-local option is
needed.

-- 
hundred-and-one symptoms of being an internet addict:
190. You quickly hand over your wallet, leather jacket, and car keys
     during a mugging, then proceed to beat the crap out of your
     assailant when he asks for your laptop.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

vim-ml commented:
I actually did implement it as a window-local option; see #3535. Let me
know if that looks good. :)

On Fri, Oct 12, 2018, 3:48 PM Bram Moolenaar <vim-dev-github@256bit.org>
wrote:

>
> > How about I do it like I said before, calling it `diffref` instead of
> > `diffmaster` and having it be on by default and settable on multiple
> > buffers at once, but then I also add `diffmaster` as a command that
> > turns it on for the current buffer and off for all others? That way it's
> > just as easy as before to use it for that most common use case of only
> > comparing with a single buffer, but at the same time if someone does
> > ever get the need to turn it on for multiple (but not all) buffers, they
> > can do that too? From my understanding of the way it's coded, it would
> > be no more difficult than only allowing a single master (actually less
> > difficult because I won't need to have it check other buffers when you
> > set the variable) so it would be a shame to explicitly remove that
> > functionality if there's another way to get the ease of use it would
> > bring.
>
> I have a hard time imagining how it would be used, have 'diffref' in
> one window/buffer to be different from another one. What are you
> looking at then?
>
> > Also you know the code a lot more than I do, but I'm pretty sure it
> > actually would be a buffer option, not a window option. It's the
> > buffer that it compares against. If two windows have the same buffer,
> > what difference would it make which of the two is set as the master?
> > The `diff` option is window-specific because it's specific to the way
> > that window is treated, but the `diffref` or `diffmaster` option,
> > whatever it's called, doesn't really involve the window; it just tells
> > other windows whether or not to reference the buffer inside.
>
> One of the most common things to do is the disable 'diff' and switch it
> on in another window. What would then happen with the buffer-local
> option? Having 'diffref' set in a buffer that doesn't appear in any
> window with 'diff' set is like a misconfiguration. We should try to
> avoid that from happening. Also, setting it for a hidden buffer doesn't
> make sense.
>
> It's true that setting the option really means to use the buffer in that
> window as the master. I don't see a problem with that.
>
> Gary's remark also mentions a situation where a window-local option is
> needed.
>
> --
> hundred-and-one symptoms of being an internet addict:
> 190. You quickly hand over your wallet, leather jacket, and car keys
> during a mugging, then proceed to beat the crap out of your
> assailant when he asks for your laptop.
>
> /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net \\\
> /// sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
> \\\ an exciting new programming language -- http://www.Zimbu.org ///
> \\\ help me help AIDS victims -- http://ICCF-Holland.org ///
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/vim/vim/pull/3527#issuecomment-429442676>
>
> --
> --
> You received this message from the "vim_dev" maillist.
> Do not top-post! Type your reply below the text you are replying to.
> For more information, visit http://www.vim.org/maillist.php
>
> ---
> You received this message because you are subscribed to a topic in the
> Google Groups "vim_dev" group.
> To unsubscribe from this topic, visit
> https://groups.google.com/d/topic/vim_dev/9CwgjallJDg/unsubscribe.
> To unsubscribe from this group and all its topics, send an email to
> vim_dev+unsubscribe@googlegroups.com.
> For more options, visit https://groups.google.com/d/optout.
>
